### DISCUSSION TITLE
There is a defect in the ref in proxyWithHistory

### DISCUSSION THREAD

**Discussion by 18613109040** [pmndrs/valtio#725] [General]
An error was reported when a ref property was recovered using the proxyWithHistory component

Please include a minimal reproduction.
For details, go to add -> undo -> redo
- [x] I've added a link to a  [codesandbox](https://codesandbox.io/s/old-shadow-v4sysy?file=/src/App.js) with a minimal reproduction.

**Comment by 18613109040** [comment]
deepClone in proxyWithHistory wants keywords for ref not to be converted to a proxy structure

```ts
const deepClone = <T>(obj: T): T => {
  if (!isObject(obj)) {
    return obj
  }
  const baseObject: T = Array.isArray(obj)
    ? []
    : Object.create(Object.getPrototypeOf(obj))
  Reflect.ownKeys(obj).forEach((key) => {
  if (key === 'ref') {
        baseObject[key as keyof T] = Reflect.get(obj, key);
      } else {
        baseObject[key as keyof T] = deepClone(obj[key as keyof T]);
      }
  })
  return baseObject
}
```

**Reply by 18613109040** [reply]
proxyWithHistory(appPages, { ignoreAttribute: "ref"})  Ignore attribute

**Comment by dai-shi** [comment]
Yeah, we found proxyWithHistory doesn't cover all cases.
Please feel free to investigate it.
You may want to create a new package if you have completely a new idea of the implementation.

**Reply by 18613109040** [reply]
When will the new package be available?

**Reply by dai-shi** [reply]
Sorry for the confusion. I modified "We" -> "You".

**Comment by dai-shi** [comment]
> An error was reported when a ref property was recovered using the proxyWithHistory component
> 
> For details, go to add -> undo -> redo

Not sure I get it. Can you reproduce it with very simple setup? Creating a repro with https://codesandbox.io/s/react-new would be good.

**Reply by 18613109040** [reply]
https://codesandbox.io/s/old-shadow-v4sysy?file=/src/App.js

**Reply by dai-shi** [reply]
Thanks. I get your point. It's about valtio `ref`.
We need ref-aware deepClone.

For now, here's the userland solution: https://codesandbox.io/s/smoosh-frost-7pd201?file=/src/App.js

**Reply by dai-shi** [reply]
#727
